### ISSUE TITLE
Metrics upload failing because an existing lock file.

### ISSUE THREAD

**Issue by mdebbar** [flutter/flutter#120440]
* https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_samsung_s10%20cull_opacity_perf__timeline_summary/5436/overview
* https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_samsung_s10%20cull_opacity_perf__timeline_summary/5437/overview
* https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_samsung_s10%20cull_opacity_perf__timeline_summary/5438/overview

**Comment by drewroengoogle** [comment]
Unable to replicate on linux-20 because local `led` runs don't upload metrics

**Comment by godofredoc** [comment]
The problem here is not the phone but somehow the auth token is invalid when trying to auth to upload the metrics file to GCS.

**Comment by godofredoc** [comment]
```
The lock is waiting for a long time: 0:43:41.440000. If the lock file flutter-flutter/2023/02/10/01/001c4951b61957f94d66fe1600efe6e0aa40248b/cull_opacity_perf__timeline_summary_intel_linux_SMG973U1_values.json.lock in bucket flutter-skia-perf-prod seems to be stuck (i.e., it was created a long time ago and no one seems to be owning it currently), delete it manually to unblock this.
Access was denied (www-authenticate header was: Bearer realm="https://accounts.google.com//", error=invalid_token).
```

**Comment by mdebbar** [comment]
Raising this to P0 since it's blocking the tree and is still failing after 4 retries.

**Comment by drewroengoogle** [comment]
@godofredoc It looks like the [first failed run](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8789588341642935569/+/u/Upload_metrics/upload_results/stdout?format=raw) provided a different error:
```
Benchmark tags to upload to skia perf: {arch: intel, device_type: SM-G973U1, device_version: none, host_type: linux}
DetailedApiRequestError(status: 504, message: )
```

**Comment by yusuf-goog** [comment]
Yes, those 2 devices (I think the S10) are experimental, and not really expected to run prod workloads.

**Comment by drewroengoogle** [comment]
Removed the hanging lock (short term solution that @godofredoc mentioned), reran the job and it passed

**Comment by drewroengoogle** [comment]
Updating issue to P3 given the short term solution. Next we need to implement the long term solution to prevent this occurring again.

**Comment by godofredoc** [comment]
Some questions:

* What caused the lingering lock file?
* Where are the configs(project, bucket, etc) defined?
* Do we have these short term solution in the LUCI playbook?

**Comment by keyonghan** [comment]
Metrics upload is handled by the metrics_center packages: https://github.com/flutter/packages/tree/main/packages/metrics_center. The immediate issue has been resolved, moving to technical debt.

@drewroengoogle Do you think you can help add the solution to the LUCI playbook?

**Comment by eyebrowsoffire** [comment]
This happened a couple of times in the last few days and I kept having to manually delete the lockfile in order to get the tree back to green.

One such example: we get a 503 here https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20opacity_peephole_col_of_rows_perf__e2e_summary/8120/overview

That failure doesn't seem to clean up the lockfile, since the subsequent attempts fail stalling on the lockfile forever: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20opacity_peephole_col_of_rows_perf__e2e_summary/8121/overview

I manually deleted the lockfile in order to get unstuck. This happened twice this weekend, breaking the tree. Respectfully, I think this should probably be higher priority than P3 since it is causing regular tree breakages.

**Comment by drewroengoogle** [comment]
> This happened a couple of times in the last few days and I kept having to manually delete the lockfile in order to get the tree back to green.
> 
> One such example: we get a 503 here https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20opacity_peephole_col_of_rows_perf__e2e_summary/8120/overview
> 
> That failure doesn't seem to clean up the lockfile, since the subsequent attempts fail stalling on the lockfile forever: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20opacity_peephole_col_of_rows_perf__e2e_summary/8121/overview
> 
> I manually deleted the lockfile in order to get unstuck. This happened twice this weekend, breaking the tree. Respectfully, I think this should probably be higher priority than P3 since it is causing regular tree breakages.

I get where you're coming from. From my understanding, P0 and P1 are more for issues that are actively blocking the tree, and while this causes breakages, there is currently a workaround (although it is annoying). Fortunately, this bug is part of [Engprod's Fixit](https://github.com/orgs/flutter/projects/37) next week, so we will be looking at it very soon. I will upgrade to P2 in the meantime though, given the more common breakages.

**Comment by drewroengoogle** [comment]
The `metrics_center` package has been updated to retry removing the lock if it fails to remove the first time.

https://github.com/flutter/flutter/pull/129788 is open, which will update to the latest metrics_center version in flutter/flutter, but it is currently blocked on some failing tests that are being addressed.

**Comment by drewroengoogle** [comment]
`metrics_center` has been updated in flutter/flutter, which implements retries for GCS errors as our first attempt to resolve the lock file issue. Please update this issue if you notice the problem still happening, and I can investigate further.

**Comment by keyonghan** [comment]
Does this https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20ios_app_with_extensions_test/11772/overview happen before or after the new fix?

**Comment by drewroengoogle** [comment]
That seems to be different, as the initial issue was a 504 instead, and when the error we are running into in this issue occurs, it fails to upload to Skia, where in the example you show, it is successfully uploading to Skia.

**Comment by drewroengoogle** [comment]
The retry was added to the lock file removal. In this case, it looks like the [retry that followed was successful](https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20ios_app_with_extensions_test/11773/overview) without having to manually delete the lock file, so I don't know if we need to add it given builds following the first failure didn't continue to fail.

**Comment by keyonghan** [comment]
Ah, I see the problem. The error is due to cocoon server side, instead of the GCS. This should be due to the daily cocoon deployment, when the server becomes unavailable for a short period of time.

Ignore my noise here!

**Comment by keyonghan** [comment]
The problem seems with `objects.insert`: https://github.com/flutter/packages/blob/main/packages/metrics_center/lib/src/gcs_lock.dart#L53

Somehow the .lock file is created, but `_api.objects.insert` hits some exception. It then keeps looping with `locked = false`.
Could we try adding a condition check before `_api.objects.insert`? If there is already a lock file, we skip insert and directly flip the status of `locked`.

> @keyonghan What are the risks associated with deleting the lock file automatically at the point where the delete it manually to unblock this message is printed? It seems like if we can tell the lock file is old, we could just remove it

By deleting the file in the `_lock` functionality, I am a bit concerned about some death lock like repeating create/delete.

**Comment by keyonghan** [comment]
what do you mean by a threshold? If the lock file exists, `_lock` will just exit and proceed to run the functionality.
